From 09c5bf534e0a3ff85fdde60e761a83e2a0fda7a9 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 27 Jun 2025 16:18:18 +0200
Subject: [PATCH 0974/1013] evl/net: ipv4: select direct routes by default

Invert the current logic which accepts any path the routing system may
find when soliciting a peer, including through gateways. Instead,
always prefer on-link routing unless EVL_NEIGH_MAYROUTE is given in
the solicitation request.

In the same move, interpret MSG_DONTROUTE in the UDP sendmsg() flags
as a request to verify that our destination is at most one hop away
from the local host (i.e. no gateway).

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/uapi/evl/net/socket-abi.h |  3 ++-
 kernel/evl/net/ipv4/ipv4.c        |  9 +++++---
 kernel/evl/net/ipv4/udp.c         | 35 ++++++++++++++++++++++++-------
 3 files changed, 36 insertions(+), 11 deletions(-)

diff --git a/include/uapi/evl/net/socket-abi.h b/include/uapi/evl/net/socket-abi.h
index e53aaaceefae..0fab7c8a798a 100644
--- a/include/uapi/evl/net/socket-abi.h
+++ b/include/uapi/evl/net/socket-abi.h
@@ -46,7 +46,8 @@ struct user_oob_msghdr {
 	struct __evl_timespec timeout;
 };
 
-#define EVL_NEIGH_PERMANENT  0x1
+#define EVL_NEIGH_PERMANENT	0x1 /* Make ARP entry permanent */
+#define EVL_NEIGH_MAYROUTE	0x2 /* Allow gateways to solicited destinations */
 
 struct evl_net_solicit {
 	struct __kernel_sockaddr_storage addr;
diff --git a/kernel/evl/net/ipv4/ipv4.c b/kernel/evl/net/ipv4/ipv4.c
index 76510ba9880c..ff853674957c 100644
--- a/kernel/evl/net/ipv4/ipv4.c
+++ b/kernel/evl/net/ipv4/ipv4.c
@@ -239,7 +239,8 @@ void __evl_net_ipv4_gc(struct evl_net_frag_tdir *ftdir)
  *		must be enabled as an oob port.
  * @addr	IPv4 address of the peer to solicit.
  * @flags	Operation flags. EVL_NEIGH_PERMANENT locks the ARP entry
- *		in the in-band cache.
+ *		in the in-band cache. EVL_NEIGH_MAYROUTE allows gateways
+ *              on the path to destination.
  *
  * Returns zero on success.
  */
@@ -252,11 +253,12 @@ int evl_net_ipv4_solicit(struct net *net,
 	struct rtable *rt;
 	__be32 ipaddr;
 	long ret = 0;
+	u8 scope;
 
 	if (addr->sa_family != AF_INET)
 		return -EAFNOSUPPORT;
 
-	if (flags & ~EVL_NEIGH_PERMANENT)
+	if (flags & ~(EVL_NEIGH_PERMANENT|EVL_NEIGH_MAYROUTE))
 		return -EINVAL;
 
 	ipaddr = ((struct sockaddr_in *)addr)->sin_addr.s_addr;
@@ -267,8 +269,9 @@ int evl_net_ipv4_solicit(struct net *net,
 	 * the routing data, which we then index into our oob route
 	 * cache.
 	 */
+	scope = flags & EVL_NEIGH_MAYROUTE ? RT_SCOPE_UNIVERSE : RT_SCOPE_LINK;
 	rt = ip_route_output(net, ipaddr, INADDR_ANY, 0,
-			     dev ? dev->ifindex : 0, RT_SCOPE_UNIVERSE);
+			     dev ? dev->ifindex : 0, scope);
 	if (IS_ERR(rt))
 		return PTR_ERR(rt);
 
diff --git a/kernel/evl/net/ipv4/udp.c b/kernel/evl/net/ipv4/udp.c
index a78e86ba1e54..f629663700ea 100644
--- a/kernel/evl/net/ipv4/udp.c
+++ b/kernel/evl/net/ipv4/udp.c
@@ -213,27 +213,39 @@ static ssize_t offload_send_udp(struct evl_socket *esk,
  * address of this peer, then the caller will have to pass on the
  * datagram to the in-band stack.
  */
-static bool find_egress_path(struct evl_socket *esk, __be32 daddr,
+static int find_egress_path(struct evl_socket *esk,
+			__be32 daddr,
 			struct evl_net_route **ertp,
 			struct evl_net_arp_entry **earpp,
-			struct evl_net_arp_entry *pseudo_earp)
+			struct evl_net_arp_entry *pseudo_earp,
+			int msg_flags)
 {
 	struct evl_net_arp_entry *earp;
 	struct evl_net_route *ert;
+	int ret = -ENOENT;
 
 	ert = evl_net_route_ipv4_output(sock_net(esk->sk), daddr);
 	if (likely(ert)) {
+		/*
+		 *  If MSG_DONTROUTE was given, make sure the
+		 *  destination is no more than one hop away from us.
+		 */
+		if  (msg_flags & MSG_DONTROUTE && rt_nexthop(ert->rt, daddr) != daddr) {
+			ret = -EMULTIHOP;
+			goto ignore;
+		}
 		earp = evl_net_get_arp_entry_or_pseudo(ert->rt->dst.dev, daddr,
 						pseudo_earp);
 		if (likely(earp))  {
 			*ertp = ert;
 			*earpp = earp;
-			return true;
+			return 0;
 		}
+	ignore:
 		evl_net_put_route(ert);
 	}
 
-	return false;
+	return ret;
 }
 
 /*
@@ -318,9 +330,14 @@ static ssize_t send_udp(struct evl_socket *esk,
 
 	/*
 	 * Unlike BSD, we accept MSG_DONTWAIT to decline waiting on
-	 * skb contention, or offloading to the in-band stage.
+	 * skb contention, or offloading to the in-band stage.  We
+	 * interpret MSG_DONTROUTE in a slightly awkward manner, in
+	 * that it makes us verify that the destination is directly
+	 * reachable from this host (i.e. on-link TOS). IOW, it does
+	 * not affect routing, it ensures that the routing information
+	 * we use is right for the task.
 	 */
-	if (msg_flags & ~MSG_DONTWAIT)
+	if (msg_flags & ~(MSG_DONTWAIT|MSG_DONTROUTE))
 		return -EINVAL;
 
 	if (evl_socket_f_flags(esk) & O_NONBLOCK)
@@ -375,7 +392,11 @@ static ssize_t send_udp(struct evl_socket *esk,
 	 * none, then offload the packet to the inband stack (as a
 	 * result, we may receive the missing information eventually).
 	 */
-	if (!find_egress_path(esk, daddr, &ert, &earp, &pseudo_earp)) {
+	ret = find_egress_path(esk, daddr, &ert, &earp, &pseudo_earp, msg_flags);
+	if (ret == -EMULTIHOP)
+		return ret;	/* MSG_DONTROUTE cannot be honored. */
+
+	if (ret) { /* No route known from the front cache - bummer. */
 		/*
 		 * We always charge the socket even when offloading to
 		 * the in-band stack although we won't consume any
-- 
2.30.2

